veloxms-core/src/main/scala/edu/berkeley/veloxms/models/Model.scala

```diff
@@ -72,30 +78,62 @@ abstract class Model[T: ClassTag] extends Logging {
    * @param nextVersion
    * @return
    */
-  def retrainFeatureModelsInSpark(observations: RDD[(UserID, T, Double)], nextVersion: Version): RDD[(T, FeatureVector)]
+  def retrainFeatureModelsInSpark(
+      observations: RDD[(UserID, T, Double)],
+      nextVersion: Version): RDD[(T, FeatureVector)]
+
+  final def retrainUserWeightsInSpark(
+      itemFeatures: RDD[(T, FeatureVector)],
+      observations: RDD[(UserID, T, Double)],
+      nextVersion: Version)
+    : RDD[(UserID, WeightVector)] = {
+
+    val newAvgUser = retrainAvgUserInSpark(itemFeatures, observations)
+    averageUser.put(newAvgUser, nextVersion)
 
-  final def retrainUserWeightsInSpark(itemFeatures: RDD[(T, FeatureVector)], observations: RDD[(UserID, T, Double)]): RDD[(UserID, WeightVector)] = {
     val obs = observations.map(x => (x._2, (x._1, x._3)))
     val ratings = itemFeatures.join(obs).map(x => (x._2._2._1, x._2._1, x._2._2._2))
     UserWeightUpdateMethods.calculateMultiUserWeights(ratings, numFeatures)
   }
 
   /**
-   * Velox implemented - fetch from local Tachyon partition
+   * Calculates the optimal cold-start user model for the case when we have
+   * no training data for a user.
+   *
+   * Implementation: Take the average score for each item across the
+   * entire training data set and use that as the training data for
+   * the average user model.
+   */
+  final def retrainAvgUserInSpark(
+      itemFeatures: RDD[(T, FeatureVector)],
+      observations: RDD[(UserID, T, Double)]): WeightVector = {
+    // drop user id from observations
+    val avgObservations = observations
+        .map(o => (o._2, o._3))
+        .groupByKey()
+        .map( {case(item, scores) => (item, (scores.reduce(_+_) / scores.size))})
+        .join(itemFeatures).map({case(item, (score, features)) => (features, score)})
+        .collect()
+    UserWeightUpdateMethods.calculateSingleUserWeights(avgObservations, numFeatures)._1
+  }
+
+  /**
    *
    */
   private def getWeightVector(userId: Long, version: Version) : WeightVector = {
+    val defaultAverage = Array.fill[Double](numFeatures)(1.0)
     val result: Option[Array[Double]] = Option(userWeights.get((userId, version)))
     result match {
       case Some(u) => u
       case None => {
         logWarning(s"User weight not found, userID: $userId")
-        averageUser
+        averageUser.get(version).getOrElse(defaultAverage)
```

Comment:
I don't think it will throw an error. `SparkVersionedBroadcast` [tries to load the version](https://github.com/dcrankshaw/velox-modelserver/blob/b75fd04956b6cc6de0c99c203080ebb0f21f97c6/veloxms-core/src/main/scala/edu/berkeley/veloxms/storage/SparkVersionedBroadcast.scala#L35), and if it can't it returns `None`.